rust/trace/src/lib.rs

```diff
@@ -0,0 +1,591 @@
+    pub name: &'static str,
+    pub payload: Option<TracePayloadT>,
+    pub start_ns: u64,
+    pub end_ns: u64,
+}
+
+impl Sample {
+    fn new(name: &'static str, payload: Option<TracePayloadT>) -> Self {
+        Self {
+            sample_id: SAMPLE_COUNTER.fetch_add(1, AtomicOrdering::Relaxed),
+            name: name,
+            start_ns: time::precise_time_ns(),
+            payload: payload,
+            end_ns: 0,
+        }
+    }
+
+    fn new_instant(name: &'static str, payload: Option<TracePayloadT>) -> Self {
+        let now = time::precise_time_ns();
+        Self {
+            sample_id: SAMPLE_COUNTER.fetch_add(1, AtomicOrdering::Relaxed),
+            name: name,
+            start_ns: now,
+            payload: payload,
+            end_ns: now
+        }
+    }
+}
+
+impl PartialEq for Sample {
+    fn eq(&self, other: &Sample) -> bool {
+        self.sample_id == other.sample_id
+    }
+}
+
+impl Eq for Sample {}
+
+impl PartialOrd for Sample {
+    fn partial_cmp(&self, other: &Sample) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Sample {
+    fn cmp(&self, other: &Sample) -> cmp::Ordering {
+        self.sample_id.cmp(&other.sample_id)
+    }
+}
+
+impl Hash for Sample {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.sample_id.hash(state);
+    }
+}
+
+pub struct SampleGuard {
+    sample: Option<Sample>,
+}
+
+impl SampleGuard {
+    #[inline]
+    fn new_disabled() -> Self {
+        Self {
+            sample: None
+        }
+    }
+
+    #[inline]
+    fn new(name: &'static str, payload: Option<TracePayloadT>) -> Self {
+        Self {
+            sample: Some(Sample::new(name, payload))
+        }
+    }
+}
+
+impl Drop for SampleGuard {
+    fn drop(&mut self) {
+        if let Some(ref mut sample) = self.sample {
+            sample.end_ns = time::precise_time_ns();
+            record_sample(sample);
+        }
+    }
+}
+
+struct Trace {
+    enabled: AtomicBool,
+    samples: Mutex<FixedLifoDeque<Sample>>,
+}
+
+impl Trace {
+    fn new() -> Self {
+        Self {
+            enabled: AtomicBool::new(false),
+            samples: Mutex::new(FixedLifoDeque::new())
+        }
+    }
+}
+
+lazy_static! { static ref TRACE : Trace = Trace::new(); }
+
+pub fn enable_tracing() {
+    enable_tracing_with_config(&Config::default());
+}
+
+pub fn enable_tracing_with_config(config: &Config) {
+    let mut all_samples = TRACE.samples.lock().unwrap();
+    all_samples.reset_limit(config.max_samples());
+    TRACE.enabled.store(true, AtomicOrdering::Relaxed);
+}
+
+pub fn disable_tracing() {
+    let mut all_samples = TRACE.samples.lock().unwrap();
+    all_samples.reset_limit(0);
+    SAMPLE_COUNTER.store(0, AtomicOrdering::Relaxed);
+    TRACE.enabled.store(false, AtomicOrdering::Relaxed);
+}
+
+#[inline]
+fn is_enabled() -> bool {
+    TRACE.enabled.load(AtomicOrdering::Relaxed)
+}
+
+pub fn trace(name: &'static str) {
+    if !is_enabled() {
```

Comment:
As a general rule I prefer the [opposite](https://www.google.com/search?q=early+return+stackoverflow&oq=early+return+stackoverflow&aqs=chrome..69i57.5455j0j4&sourceid=chrome&ie=UTF-8) :).  It mainly makes sense in larger snippets.  I'm happy to switch this one since they're tiny & unlikely to grow further.

Reply:
Yea, one thing I like about swift is that there's an explicit `guard` statement for handling early returns.